src/com/shashi/service/impl/ProductServiceImpl.java

```diff
@@ -409,6 +410,71 @@ public List<ProductBean> getLowStockProduct() {
 		return products;
 	}
 
+	@Override
+	public List<ProductBean> getUnpopularProduct() {
+		List<ProductBean> products = new ArrayList<>();
+		int defaultThreshold = 3;
+
+		Connection con = DBUtil.provideConnection();
+
+		PreparedStatement ps = null;
+		ResultSet rs = null;
+
+        try {
+
+            // Get all products
+            List<ProductBean> allProducts = getAllProducts();
+
+            // Iterate through each product and check sold quantity
+            for (ProductBean p : allProducts) {
+                int soldQuantity = 0;
+
+				con = DBUtil.provideConnection();
+				ps = null;
```

Comment:
instead of all these queries you can do a query similar to the sortProductsBySales method. You could do a query like this (just rename to the right columns):
`SELECT p.*
FROM products p
WHERE
    p.discount IS NULL AND 
    p.productid IN (
        SELECT o.productid
        FROM orders o
        GROUP BY o.productid
        HAVING SUM(o.quantity) < default_treshold 
    );`

this should return the list of products directly in 1 query